crypto/evp/pmeth_lib.c

```diff
@@ -70,6 +70,69 @@ static pmeth_fn standard_methods[] = {
 #endif
 };
 
+/* This must be in the same order as standard_methods. */
+/* It's the same array with non-FIPS methods removed. */
+static pmeth_fn standard_methods_fips[] = {
+#ifndef OPENSSL_NO_RSA
+    rsa_pkey_method,
+#endif
+#ifndef OPENSSL_NO_DH
+    dh_pkey_method,
+#endif
+#ifndef OPENSSL_NO_DSA
+#ifndef OPENSSL_FIPS
+    dsa_pkey_method,
+#endif
+#endif
+#ifndef OPENSSL_NO_EC
+    ec_pkey_method,
+#endif
+    hmac_pkey_method,
+#ifndef OPENSSL_NO_CMAC
+    cmac_pkey_method,
+#endif
+#ifndef OPENSSL_NO_RSA
+    rsa_pss_pkey_method,
+#endif
+#ifndef OPENSSL_NO_DH
+    dhx_pkey_method,
+#endif
+#ifndef OPENSSL_NO_SCRYPT
+#ifndef OPENSSL_FIPS
+    scrypt_pkey_method,
```

Comment:
FWIW, scrypt uses FIPS-approved SHA-256 on the outer layer and its cryptographic security for the password input is guaranteed by SHA-256, HMAC, PBKDF2. I don't know if this possibly makes it FIPS-compliant despite of it also using Salsa20 internally?

I am using a similar argument to justify usage of yescrypt where NIST-approved primitives are needed - in there, we have entropy bypasses for both password and salt inputs, so cryptographic security is provably guaranteed by SHA-256, HMAC, PBKDF2, whereas the actually most important (against GPUs, etc.) internal processing is "non-cryptographic".

I think OpenSSL-provided scrypt isn't used much yet (it's a recent addition to OpenSSL), so I am more interested in whether this is a stance we'd like to defend in general. We're going to have yescrypt for user passwords on EL 10 by default.

Reply:
It's not FIPS compliant. It's not listed in our security policy, so we have to disable it.